streams/src/main/java/org/apache/kafka/streams/processor/internals/RecordCollectorImpl.java

```diff
@@ -98,16 +95,14 @@ public RecordCollectorImpl(final Producer<byte[], byte[]> producer, final String
                     @Override
                     public void onCompletion(final RecordMetadata metadata, final Exception exception) {
                         if (exception == null) {
-                            if (sendException != null) {
-                                return;
-                            }
                             final TopicPartition tp = new TopicPartition(metadata.topic(), metadata.partition());
                             offsets.put(tp, metadata.offset());
                         } else {
-                            if (sendException == null) {
-                                sendException = exception;
-                                log.error("{} Error sending record to topic {}. No more offsets will be recorded for this task and the exception will eventually be thrown", logPrefix, topic, exception);
-                            }
+                            log.error("{} Error sending record with key {} to topic {}, partition {}. " +
+                                    "No more offsets will be " +
+                                    "recorded for this task and the exception will eventually be thrown",
+                                logPrefix, key, topic, partition, exception);
+                            throw new StreamsException(String.format("%s exception caught when producing", logPrefix), exception);
```

Comment:
Thanks @dguy you're right. @mjsax you recently looked at exceptions, what do you recommend? We only know if something went wrong in the callback at some point in the future. Meanwhile we could be sending other requests (with the previous code we checked if the last request gave an exception, however the check can't guarantee much since the last request might have not completed yet as we're sending the new one). From an EOS perspective where do you want to see the exception? Thanks. cc @guozhangwang